src/transaction_builder.cpp

```diff
@@ -0,0 +1,281 @@
+
+#include <boost/variant.hpp>
+#include <librustzcash.h>
+
+SpendDescriptionInfo::SpendDescriptionInfo(
+    libzcash::SaplingExpandedSpendingKey xsk,
+    libzcash::SaplingNote note,
+    uint256 anchor,
+    ZCSaplingIncrementalWitness witness) : xsk(xsk), note(note), anchor(anchor), witness(witness)
+{
+    librustzcash_sapling_generate_r(alpha.begin());
+}
+
+TransactionBuilder::TransactionBuilder(
+    const Consensus::Params& consensusParams,
+    int nHeight,
+    CKeyStore* keystore) : consensusParams(consensusParams), nHeight(nHeight), keystore(keystore)
+{
+    mtx = CreateNewContextualCMutableTransaction(consensusParams, nHeight);
+}
+
+bool TransactionBuilder::AddSaplingSpend(
+    libzcash::SaplingExpandedSpendingKey xsk,
+    libzcash::SaplingNote note,
+    uint256 anchor,
+    ZCSaplingIncrementalWitness witness)
+{
+    // Consistency check: all anchors must equal the first one
+    if (!spends.empty()) {
+        if (spends[0].anchor != anchor) {
+            return false;
+        }
+    }
+
+    spends.emplace_back(xsk, note, anchor, witness);
+    mtx.valueBalance += note.value();
+    return true;
+}
+
+void TransactionBuilder::AddSaplingOutput(
+    libzcash::SaplingFullViewingKey from,
+    libzcash::SaplingPaymentAddress to,
+    CAmount value,
+    std::array<unsigned char, ZC_MEMO_SIZE> memo)
+{
+    auto note = libzcash::SaplingNote(to, value);
+    outputs.emplace_back(from.ovk, note, memo);
+    mtx.valueBalance -= value;
+}
+
+void TransactionBuilder::AddTransparentInput(COutPoint utxo, CScript scriptPubKey, CAmount value)
+{
+    if (keystore == nullptr) {
+        throw std::runtime_error("Cannot add transparent inputs to a TransactionBuilder without a keystore");
+    }
+
+    mtx.vin.emplace_back(utxo);
+    tIns.emplace_back(scriptPubKey, value);
+}
+
+bool TransactionBuilder::AddTransparentOutput(CTxDestination& to, CAmount value)
+{
+    if (!IsValidDestination(to)) {
+        return false;
+    }
+
+    CScript scriptPubKey = GetScriptForDestination(to);
+    CTxOut out(value, scriptPubKey);
+    mtx.vout.push_back(out);
+    return true;
+}
+
+void TransactionBuilder::SetFee(CAmount fee)
+{
+    this->fee = fee;
+}
+
+void TransactionBuilder::SendChangeTo(libzcash::SaplingPaymentAddress changeAddr, libzcash::SaplingFullViewingKey fvkOut)
+{
+    zChangeAddr = std::make_pair(fvkOut, changeAddr);
+    tChangeAddr = boost::none;
+}
+
+bool TransactionBuilder::SendChangeTo(CTxDestination& changeAddr)
+{
+    if (!IsValidDestination(changeAddr)) {
+        return false;
+    }
+
+    tChangeAddr = changeAddr;
+    zChangeAddr = boost::none;
+}
+
+boost::optional<CTransaction> TransactionBuilder::Build()
+{
+    //
+    // Consistency checks
+    //
+
+    // Valid change
+    CAmount change = mtx.valueBalance - fee;
+    for (auto tIn : tIns) {
+        change += tIn.value;
+    }
+    for (auto tOut : mtx.vout) {
+        change -= tOut.nValue;
+    }
+    if (change < 0) {
+        return boost::none;
```

Comment:
Maybe we should return a value that indicates why building the tx failed?

Reply:
Were this Rust, I'd be returning a `Result<CTransaction, Reason>`. The equivalent in C++ is very messy; probably the simplest is to return a `std::pair<boost::optional<CTransaction>, Reason>`, or we could add a `GetFailureReason()` method that exposes an internal variable we set before returning from `Build()`.

Reply:
I would be inclined to implement a `Result` class.